### PR TITLE
Pz/types 010 - loyalties members

### PR DESCRIPTION
**Pull request by p-zielinski**
---
'@voucherify/sdk': minor
---

Add support for few endpoints of Loyalties API.
- Added support for new endpoints: `GET /loyalties/members/{memberId}`, `GET /loyalties/members/{memberId}/activities`, `GET /loyalties/members/{memberId}/rewards`, `POST /loyalties/{campaignId}/members/{memberId}/transfers`, `GET /loyalties/{campaignId}/members/{memberId}/points-expiration`, `GET /loyalties/members/{memberId}/transactions`, `GET /loyalties/{campaignId}/members/{memberId}/transactions`, `POST /loyalties/members/{memberId}/transactions/export` and `POST /loyalties/{campaignId}/members/{memberId}/transactions/export` [(examples of usage available in readme.md)](..%2F..%2Fpackages%2Fsdk%2FREADME.md)
- New exported types/interfaces: `LoyaltiesTransferPointsResponseBody`, `LoyaltiesTransferPointsRequestBody`, `LoyaltiesListMemberRewardsRequestQuery`, `LoyaltiesListMemberRewardsResponseBody`, `LoyaltiesGetPointsExpirationRequestQuery`, `LoyaltiesGetPointsExpirationResponseBody`, `LoyaltiesListCardTransactionsRequestQuery`, `LoyaltiesListCardTransactionsResponseBody`, `LoyaltiesExportCardTransactionsRequestBody`, `LoyaltiesExportCardTransactionsResponseBody`, `LoyaltiesAddOrRemoveCardBalanceRequestBody`, `LoyaltiesAddOrRemoveCardBalanceResponseBody`, `LoyaltyCardTransaction`, `SimpleLoyaltyVoucher`, `LoyaltiesTransferPoints`, `LoyaltyCardTransactionsFields`, `LoyaltyCardTransactionsType`, `Reward`, `RewardTypeCampaign`, `RewardTypeCoin`, `RewardTypeMaterial`, `RewardType`, `RewardAssignment`

### CODE REVIEW

File: packages/sdk/README.md [hunk 1]
```diff
@@ -1064,6 +1072,12 @@ client.loyalties.createMember(campaignId, member)
 client.loyalties.getMember(campaignId, memberId)
 ```
 
+#### [Get Loyalty Program Member Without CampaignId](https://docs.voucherify.io/reference/get-member-1)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1075]: Few issues here:
1. The header and method name suggest that we do not need a campaign ID, but you expect it in params and link to documentation where the campaign ID is required.
2. I do not like naming. I understand that we have two similar endpoints, one that expects just member ID and the second one that needs campaign ID in addition. In documentation, both are described as `Get Member`. I guess (but we can confirm it) that the one that needs a campaign ID is a legacy one. We should give a name that will be intuitive for a developer who does not know about the context (legacy method). So, maybe `getMemberById`?

File: packages/sdk/README.md [hunk 2]
```diff
@@ -1079,14 +1093,52 @@ client.loyalties.getMemberActivities(campaignId, memberId)
...
+```
+
+`memberId` referrers to Loyalty Card code.
+
+---
+
+#### [Add Or Remove Loyalty Card Balance](https://docs.voucherify.io/reference/add-remove-loyalty-card-balance-1)
+
+```javascript
+client.loyalties.addOrRemoveLoyaltyCardBalance(memberId, balance, campaignId)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1117]: Naming: `addOrRemoveLoyaltyCardBalance` -> `addOrRemoveCardBalance`?

File: packages/sdk/README.md [hunk 3]
```diff
@@ -1099,6 +1151,46 @@ When redeeming reward with type `COIN` you need to provide additional `order` ob
 
 ---
 
+#### [List Loyalty Card Transactions](https://docs.voucherify.io/reference/list-loyalty-card-transactions)
+
+```javascript
+client.loyalties.listLoyaltyCardTransactions(memberId)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1157]: `listLoyaltyCardTransactions` -> `listCardTransactions`?

File: packages/sdk/README.md [hunk 4]
```diff
@@ -1099,6 +1151,46 @@ When redeeming reward with type `COIN` you need to provide additional `order` ob
...
+```
+
+`memberId` referrers to Loyalty Card code.
+
+---
+
+#### [List Loyalty Card Transactions With CampaignId](https://docs.voucherify.io/reference/list-loyalty-card-transactions-1)
+
+```javascript
+client.loyalties.listLoyaltyCardTransactionsWithCampaignId(campaignId, memberId)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1167]: `listLoyaltyCardTransactionsWithCampaignId` -> listTransactionsWithCampaignId`?

File: packages/sdk/README.md [hunk 5]
```diff
@@ -1099,6 +1151,46 @@ When redeeming reward with type `COIN` you need to provide additional `order` ob
...
+```
+
+`memberId` referrers to Loyalty Card code.
+
+---
+
+#### [Export Loyalty Card Transactions](https://docs.voucherify.io/reference/export-loyalty-card-transactions)
+
+```javascript
+client.loyalties.exportLoyaltyCardTransactions(memberId, exportLoyaltyCardTransactionsParams)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1177]: `exportLoyaltyCardTransactions` -> `exportCardTransactions`?

File: packages/sdk/README.md [hunk 6]
```diff
@@ -1099,6 +1151,46 @@ When redeeming reward with type `COIN` you need to provide additional `order` ob
...
+```
+
+`memberId` referrers to Loyalty Card code.
+
+---
+
+#### [Export Loyalty Card Transactions With CampaignId](https://docs.voucherify.io/reference/export-loyalty-card-transactions-1)
+
+```javascript
+client.loyalties.exportLoyaltyCardTransactionsWithCampaignId(campaignId, memberId, exportLoyaltyCardTransactionsParams)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1187]: `exportLoyaltyCardTransactionsWithCampaignId` -> `exportCardTransactionsWithCampaignId`?

**Review by marcin-slezak — CHANGES_REQUESTED** [review]
Adjust type names to a new guideline

File: packages/sdk/README.md [hunk 7]
```diff
@@ -966,10 +966,17 @@ Methods are provided within `client.loyalties.*` namespace.
...
 - [Create Loyalty Program Member](#create-loyalty-program-member)
 - [Get Loyalty Program Member](#get-loyalty-program-member)
-- [List Loyalty Program Members](#list-loyalty-members)
+- [Get Loyalty Program Member By Id](#get-loyalty-program-member-by-id)
+- [List Loyalty Program Members](#list-loyalty-program-members)
 - [Get Loyalty Program Member Activities](#get-loyalty-program-member-activities)
-- [Add Loyalty Card Balance](#add-loyalty-card-balance)
+- [Get Loyalty Program Member Activities Without CampaignId](#get-loyalty-program-member-activities-without-campaignid)
+- [Add Or Remove Loyalty Card Balance](#add-or-remove-loyalty-card-balance)
+- [[Deprecated] Add Loyalty Card Balance](#add-loyalty-card-balance)
```
Comment (p-zielinski) [inline packages/sdk/README.md:974]: added

File: packages/sdk/README.md [hunk 8]
```diff
@@ -1072,21 +1078,62 @@ client.loyalties.listMembers(campaignId, params)
...
+```javascript
+client.loyalties.listMemberRewards(memberId, params)
+```
+
+`memberId` referrers to Loyalty Card code.
+
+---
+
+#### [Add Or Remove Loyalty Card Balance](https://docs.voucherify.io/reference/add-remove-loyalty-card-balance)
+#### [Add Or Remove Loyalty Card Balance 2](https://docs.voucherify.io/reference/add-remove-loyalty-card-balance-1)
```
Comment (marcin-slezak) [inline packages/sdk/README.md:1101]: Do not duplicate headers if there are two pages in documentation. Maybe:

```markdown
### [Some header](https://first-link.com)

Depending on the parameters, this method sends requests to [first](https://first-link.com) or [second](https://second-link.com) API endpoint
```

File: packages/sdk/src/types/Loyalties.ts [hunk 9]
```diff
@@ -497,3 +499,222 @@ export interface LoyaltyPointsTransfer {
 	code: string
 	points: number
 }
+
+// 0-level types
+
+export type LoyaltiesTransferPointsResponseBody = LoyaltiesVoucherResponse
```
Comment (marcin-slezak) [inline packages/sdk/src/types/Loyalties.ts:505]: try to avoid references to legacy types

File: packages/sdk/src/types/Loyalties.ts [hunk 10]
```diff
@@ -497,3 +499,222 @@ export interface LoyaltyPointsTransfer {
...
+export interface LoyaltiesListMemberRewardsRequestQuery {
+	affordable_only?: boolean
+	limit?: number
+	page?: number
+}
+
+export interface LoyaltiesListMemberRewardsResponseBody {
+	object: 'list'
+	data_ref: 'data'
+	data: { reward: RewardsGetResponse; assignment: RewardsAssignmentObject; object: 'loyalty_reward' }[]
```
Comment (marcin-slezak) [inline packages/sdk/src/types/Loyalties.ts:518]: formatting, reference to legacy types

**Review by marcin-slezak — CHANGES_REQUESTED** [review]

**Review by marcin-slezak — CHANGES_REQUESTED** [review]

**Review by marcin-slezak — APPROVED** [review]

**Pull request merged**